proposals/NNNN-unlock-existential-types-for-all-protocols.md

```diff
@@ -0,0 +1,208 @@
+# Unlock Existentials for All Protocols
+
+* Proposal: [SE-NNNN](NNNN-unlock-existential-types-for-all-protocols.md)
+* Authors: [Anthony Latsis](https://github.com/AnthonyLatsis), [Filip Sakel](https://github.com/filip-sakel), [Suyash Srijan](https://github.com/theblixguy)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#33767](https://github.com/apple/swift/pull/33767)
+
+
+### Introduction
+
+Swift currently offers the ability for protocols that meet certain criteria to be used as types. Trying to use an unsupported protocol as a type yields the error `[the protocol] can only be used as a generic constraint because it has 'Self' or associated type requirements`. This proposal aims to relax this artificial constraint imposed on such protocols.
+
+
+### Motivation
+
+Currently, any protocol that has [non-covariant](https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)) `Self` references or associated type requirements is not allowed to be used as a type. Initially, this restriction reflected technical limitations (as Joe states [here](https://forums.swift.org/t/lifting-the-self-or-associated-type-constraint-on-existentials/18025)); however, such limitations have now been alleviated. As a result, users are left unable to utilize a powerful feature for certain protocols. That’s evident in a plethora of projects. For instance, the Standard Library has existential types such as [`AnyHashable`](https://developer.apple.com/documentation/swift/anyhashable) and [`AnyCollection`](https://developer.apple.com/documentation/swift/anycollection) and SwiftUI has [`AnyView`](https://developer.apple.com/documentation/swiftui/anyview). 
+
+Generics are most often the best mechanism for type-level abstraction, which relies on the compiler knowing type information during compilation. However, type information is not always a gurantee, which is why the value-level abstraction of existential types is extremely useful in some cases.
+
+One such case is heterogenous collections, which require value-level abstraction to store their elements of various types:
+
+```swift
+protocol Identifiable {
+    associatedtype ID: Hashable 
+    
+    var id: ID { get }
+}
```

Comment:
It might be good to expand here with a code sample that shows creating an array of Identifable leads to an error, which is why we need to create a type eraser wrapper.

Reply:
I changed it to this:

```swift
let naiveIdentifiables: [Identifiable] ❌
// The compiler doesn't currently allow that.
// So, we work around that by creating a
// custom existential type: `AnyIdentifiable`.
```

Let me know if you think it's too detailed or whether you think we should add a more detailed message. I chose to not expand too much on the actual error message, since SE readers are probably familiar with that. Additionally, since I've added this message, I've also added the reason why we need he existential - as you proposed.

Reply:
Looks good. One nit though - we should use `'` instead of `` ` `` or just drop it entirely since backtracks don't work in comments.